crates/lang/macro/src/lib.rs

```diff
@@ -18,20 +18,600 @@ mod trait_def;
+///     #[ink::contract(compile_as_dependency = true)]
+///     mod my_contract {
+///         # #[ink(storage)]
+///         # pub struct MyStorage;
+///         # impl MyStorage {
+///         #     #[ink(constructor)]
+///         #     pub fn construct() -> Self { MyStorage {} }
+///         #     #[ink(message)]
+///         #     pub fn message(&self) {}
+///         # }
+///         // ...
+///     }
+///     ```
+///
+///     **Default value:** Depends on the crate feature propagation of `Cargo.toml`.
+///
+/// - `env: impl EnvTypes`
+///
+///     Tells the ink! code generator which environment to use for the ink! smart contract.
+///     The environment must implement the `EnvTypes` (defined in `ink_env`) trait and provides
+///     all the necessary fundamental type definitions for `Balance`, `AccountId` etc.
+///
+///     **Usage Example:**
+///
+///     Given a custom `EnvTypes` implementation:
+///     ```
+///     pub struct MyEnvTypes;
+///
+///     impl ink_env::EnvTypes for MyEnvTypes {
+///         const MAX_EVENT_TOPICS: usize = 3;
+///         type AccountId = u64;
+///         type Balance = u128;
+///         type Hash = [u8; 32];
+///         type Timestamp = u64;
+///         type BlockNumber = u32;
+///     }
+///     ```
+///     A user might implement their ink! smart contract using the above custom `EnvTypes`
+///     implementation as demonstrated below:
+///     ```
+///     # use ink_lang as ink;
+///     #[ink::contract(env_types = MyEnvTypes)]
+///     mod my_contract {
+///         # pub struct MyEnvTypes;
+///         #
+///         # impl ink_env::EnvTypes for MyEnvTypes {
+///         #     const MAX_EVENT_TOPICS: usize = 3;
+///         #     type AccountId = u64;
+///         #     type Balance = u128;
+///         #     type Hash = [u8; 32];
+///         #     type Timestamp = u64;
+///         #     type BlockNumber = u32;
+///         # }
+///         #
+///         # #[ink(storage)]
+///         # pub struct MyStorage;
+///         # impl MyStorage {
+///         #     #[ink(constructor)]
+///         #     pub fn construct() -> Self { MyStorage {} }
+///         #     #[ink(message)]
+///         #     pub fn message(&self) {}
+///         # }
+///         // ...
+///     }
+///     ```
+///
+///     **Default value:** `DefaultEnvTypes` defined in `ink_env` crate.
+///
+/// ## Anaylsis
+///
+/// The `#[ink::contract]` macro fully analyses its input smart contract
+/// against invalid arguments and structure.
+///
+/// Some example rules include but are not limited to:
+///
+/// - There must be exactly one `#[ink(storage)]` struct.
+///
+///     This struct defined the layout of the storage that the ink! smart contract operates on.
+///     The user is able to use a variety of built-in facitilies, combine them in various way
+///     or even provide their own implementations of storage data structures.
+///
+///     For more information the user shall visit the `ink_storage` crate documentation.
```

Comment:
```suggestion
///     For more information visit the `ink_storage` crate documentation.
```
Would be great to have a doc link here too